kernel1.ml

```diff
@@ -0,0 +1,68 @@
+(*Kernel 1 is basic construction of adjacency HashMap for undirected graphs which is corresponding to sparse graph implementation. INPUTS : ijw and m which has been 
+derived from kronecker product*)
+
+(*(*<-------OCaml Kernel 1 inspired from https://graph500.org/?page_id=12---------->
+Written by support of PRISM Lab, IIT Madras and OCaml Labs*)*)
+
+(*This function helps in transpose of the list which has to be converted from (startVertex, endVertex, weight) in column to (startVertex, endVertex, weight) in 3 rows*)
+let rec transpose list col newList = 
+	if col = 3 then newList else 
+		let rec transposeRow row rowList =
+			if row = List.length list then rowList else transposeRow (row+1) (rowList@[ List.nth (List.nth list row) col ])
+		in transpose list (col+1) (newList@[(transposeRow 0 [])])  
+;;
+
+(*This basically sorts the list in a way that (startVertex, endVertex), startVertex > endVertex.*)
+let rec sortVerticeList list newList =
+	let rec sortVerticeList list newList maximum = 
+		match list with
+		[] -> (newList,maximum) |
+		head::tail -> let x = List.nth head 0 and y = List.nth head 1 in if x > y then sortVerticeList tail (newList@[[x;y;List.nth head 2]]) (max maximum x)     
+		else sortVerticeList tail (newList@[[y;x;List.nth head 2]]) (max maximum y)
+	in sortVerticeList list newList 0
+;;
+
+(*As the name suggests, it removes the self loops from ijw*)
+let rec removeSelfLoops ijw newList col m =
+	if col = m then newList else begin
+		if  List.nth (List.nth ijw 0) col = List.nth (List.nth ijw 1) col 
+		then removeSelfLoops ijw newList (col+1) m 
+	else removeSelfLoops ijw (newList@[ [(List.nth (List.nth ijw 0) col) ; (List.nth (List.nth ijw 1) col) ; (List.nth (List.nth ijw 2) col) ] ]) (col+1) m
+	end 
+;;
+
+(*This is basically the construction of adj matrix [row][col], just in case dense graphs are being tested. All the kernels further though use HashMap, and thus would 
+require changes*)
+
+(*let constructionAdjMatrix list maxLabel =
+	 let matrix = Array.make_matrix maxLabel maxLabel 0. in
+	 let rec fillMatrix matrix list =
+	 	match list with
+	 	[] -> matrix |
+	 	head::tail -> let _ = matrix.(int_of_float(List.nth head 0)).(int_of_float(List.nth head 1)) <- (List.nth head 2) in  
+	 				let _ = matrix.(int_of_float(List.nth head 1)).(int_of_float(List.nth head 0)) <- (List.nth head 2) in fillMatrix matrix tail
+	in fillMatrix matrix list
+;;*)
+
+(*Adding Edge adds the edge to HashMap for undirected graphs, where the binding is between index and the list (endVertex, weight) *)
+
+let addEdge startVertex endVertex weight hashTable =
+	if Hashtbl.mem hashTable startVertex = false
+	then let _= Hashtbl.add hashTable startVertex [(endVertex,weight)] in hashTable
+	else let _ = Hashtbl.replace hashTable startVertex ( (Hashtbl.find hashTable startVertex) @ [(endVertex,weight)]) in hashTable
```

Comment:
If I remove it, then it shows syntax error. Also, let _ = is added because it will allow the usage of in keyword else I can use ';' if that's the general rule!

Reply:
if you remove `let _ =` you should remove the corresponding `in` also. 
```ocaml
then let _= Hashtbl.add hashTable startVertex [(endVertex,weight)] in hashTable
```
can be written as
```
Hashtbl.add hashTable startVertex [(endVertex,weight)];
hashTable
```
Also, is there a need to return the hash table here? Like I mentioned above, `Hashtbl` is a mutable data structure, so if you are going to use the same instance of a hash table, then you probably don't have to return it every time you do some addition or modification to the table.

Reply:
Done. I was confused whether it will refer to the same hashtable or not, but its like a pointer and I guess mutable means that it goes to that location and changes, instead of creating a copy which is nice.

Reply:
All right. You can probably remove `let _ =` present inside the functions elsewhere in the code, too.